Add tutorial on specifying custom gradients

sea-bass opened the pull request:
@EricCousineau-TRI recently helped me figure out how to specify your own custom gradients, both analytically and with finite differencing, so I owe you guys a tutorial.

See here for the `nbviewer` link: https://nbviewer.org/github/sea-bass/drake/blob/custom-gradients-tutorial/tutorials/custom_gradients.ipynb

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RobotLocomotion/drake/21502)
<!-- Reviewable:end -->

RussTedrake reviewed (commented):
Thanks Sebastian. I do think having something like this will be very helpful.

I've left some quick drive-by comments below.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzEmY2P-fqSbl_0nDGC:b-qu1xj3)* status: 3 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 61 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzEmpIPE9vkKVVxve50:-NzEmpIPE9vkKVVxve51:bihn4i5) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L61)):*
> ```
>     "For simple expressions like these, AutoDiff already took care of calculating all the gradients for optimization.\n",
>     "\n",
>     "However, assume you have a more complex function that cannot accept these `AutoDiffXd` object that get created by Drake. In this case, you can manually specify gradients yourself, for both cost and constraint expressions."
> ```

This is exactly right. But I'd like to be super clear here that most people do not need to do this.  We should be clear that even e.g.

```
def cost_expression(x):
    return x[0] + 2.0 * x[1]

prog.AddCost(cost_expression, x)
```

Will work fine -- and support autodiff.  As you say, it would typically only be if you're calling e.g. some black-box method that cannot consume the AutoDiffXd type that you would need to do it.
___
*[`tutorials/custom_gradients.ipynb` line 113 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzElg1u02bM9NAUunNh:-NzElg1v9nmxeK81NT1j:b-ta3z5j) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L113)):*
> ```
>     "        # Return the new AutoDiffXd object with the constraint and gradient\n",
>     "        # NOTE: In this case, we have to wrap the expression in a numpy array\n",
>     "        return np.array([AutoDiffXd(constraint, gradient)])\n",
> ```

note: I think we should recommend calling `InitializeAutoDiff(value, gradient)` here; in general constraints are vectors and gradients are (Jacobian) matrices, which could be difficult to get right.  For that reason, I would also recommend making the example constraint have num_constraints > 1.
___
*[`tutorials/custom_gradients.ipynb` line 199 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzEnjRr9f0AnbA32Uhv:-NzEnjRr9f0AnbA32Uhw:b-q40chd) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L199)):*
> ```
>     "        \n",
>     "        # Return the new AutoDiffXd object with the constraint and gradient\n",
>     "        return np.array([AutoDiffXd(constraint[0,0], gradient[0,:])])\n",
> ```

Prefer InitializeAutoDiff here, too.


<!-- Sent from Reviewable.io -->

sea-bass commented:
*[`tutorials/custom_gradients.ipynb` line 113 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzElg1u02bM9NAUunNh:-NzFeCNn6gSPf-begJ96:bptsqsf) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L113)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

note: I think we should recommend calling `InitializeAutoDiff(value, gradient)` here; in general constraints are vectors and gradients are (Jacobian) matrices, which could be difficult to get right.  For that reason, I would also recommend making the example constraint have num_constraints > 1.
</blockquote></details>

Re: `num_constraints > 1`, I was debating something like this.

Basically a tradeoff between pedagogical clarity (1 constraint) and actually motivating the need for an array of things.

I can probably do 2 "stay out of this circle" constraints which strikes a good balance.


<!-- Sent from Reviewable.io -->

sea-bass commented:
Thanks for the review, @RussTedrake! I implemented your feedback and also added plots and equations to better illustrate the problem. I'm much happier with it now.

sea-bass commented on tutorials/custom_gradients.ipynb:
```diff
@@ -0,0 +1,339 @@
+    "\n",
+    "$\\begin{aligned}\n",
+    "\\frac{\\partial \\text{constraint}}{\\partial x} = \\begin{bmatrix}\n",
+    "2x_0 & 2x_1 \\\\ 2(x_0 - 0.5) & 2(x_0 + 0.1)\n",
```
```suggestion
    "2x_0 & 2x_1 \\\\ 2(x_0 - 0.5) & 2(x_1 + 0.1)\n",

jwnimmer-tri commented:
Would any volunteers like to sign up as feature reviewer @EricCousineau-TRI @hongkai-dai @RussTedrake?

EricCousineau-TRI reviewed (commented):
Yup! +@EricCousineau-TRI - thanks for the tutorial!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzTYHhZ0xSLp2x7vv8c:b-59qo5h)* status: 4 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)



<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
First pass done, PTAL!

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzTdiT-8SoFhGkyN6ru:b-v6scke)* status: 12 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTdnwR3Xe_jWssS_6d:-NzTdnwR3Xe_jWssS_6e:b-l26meu) (no related file):*
Before merging, should clear the notebook output.

___
___
*[`tutorials/custom_gradients.ipynb` line 113 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzElg1u02bM9NAUunNh:-NzTdVzT7_CPDhJ72PJY:b-d3ades) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L113)):*
<details><summary><i>Previously, sea-bass (Sebastian Castro) wrote…</i></summary><blockquote>

Re: `num_constraints > 1`, I was debating something like this.

Basically a tradeoff between pedagogical clarity (1 constraint) and actually motivating the need for an array of things.

I can probably do 2 "stay out of this circle" constraints which strikes a good balance.
</blockquote></details>

Sorry to tack on one more thing, but just to ensure we have the shaping correct (e.g. chain rule), ideally we have non-square matrices.

Would it be possible to have one more decision variable, such that the Jacobian will be a 2x3?
___
*[`tutorials/custom_gradients.ipynb` line 131 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTYv2p7QHDJamEx4W1:-NzTYv2p7QHDJamEx4W2:b-fgirxf) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L131)):*
> ```
>     "For this specific example, we have already seen that providing gradients is not needed.\n",
>     "\n",
>     "However, if you are using more complex expressions or calling out to libraries that do not work well with `AutoDiffXd` objects, you may need to provide the gradients yourself.\n",
> ```

nit "complex expressions" may be ambiguous. Consider expanding:

Consider instead:
"However, if your computations come from a third-party library, or are complex enough that they are intractable or slow via Drake's automatic differentiation [...]"
___
*[`tutorials/custom_gradients.ipynb` line 133 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTZNrJDtWD2ylZiNC5:-NzTZNrJDtWD2ylZiNC6:bspre9g) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L133)):*
> ```
>     "However, if you are using more complex expressions or calling out to libraries that do not work well with `AutoDiffXd` objects, you may need to provide the gradients yourself.\n",
>     "\n",
>     "If possible, you should provide analytical gradients as shown below. We can do this because we know the derivatives of the costs and constraints.\n",
> ```

nit "If possible" sounds like a user should prefer this over AutoDiff.
Consider rewriting / loosening:
"For the sake of this example, we wish to manually compute the analytical gradients for the cost and constraints as follows"
___
*[`tutorials/custom_gradients.ipynb` line 178 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzT_ip2C1xHMLUcVsIy:-NzT_ip2C1xHMLUcVsIz:bwg1b9k) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L178)):*
> ```
>     "        # NOTE: x_value is a column vector so we specify two indices to avoid warnings\n",
>     "        cost = x_value[0,0] + 2.0 * x_value[1,0]\n",
>     "        gradient = np.array([1.0, 2.0])\n",
> ```

I believe this might be missing the chain rule? (`ExtractGradient`)

I believe this is currently working because the gradients supplied are currently identity.
However, if the purpose of this tutorial is the correct way to supply gradients to AutoDiff, we should ideally supply the correct value here.
___
*[`tutorials/custom_gradients.ipynb` line 199 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTaP8b8yCzno5yWFhH:-NzTaP8b8yCzno5yWFhI:b91kkqj) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L199)):*
> ```
>     "            (x_value[0, 0] - 0.5)**2 + (x_value[1, 0] + 0.1)**2,\n",
>     "        ])\n",
>     "        gradient = 2.0 * np.array([\n",
> ```

Same as above - missing chain rule?
___
*[`tutorials/custom_gradients.ipynb` line 270 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTc3cA2pJd7s52MZVE:-NzTaYldAYvgrnPc27sx:b91kkqj) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L270)):*
> ```
>     "        # We are using forward differencing here as the `option` argument.\n",
>     "        cost = cost_fn(x_value)\n",
>     "        gradient = ComputeNumericalGradient(cost_fn, x_value, option=NumericalGradientOption(NumericalGradientMethod.kForward))\n",
> ```

Same as above - missing chain rule?
___
*[`tutorials/custom_gradients.ipynb` line 284 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTbx8M7c0WWg5tiT8h:-NzTbx8M7c0WWg5tiT8i:bxt8tvg) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L284)):*
> ```
>     "    if x.dtype == float:\n",
>     "        # The constraint expression with floats\n",
>     "        return np.array([[\n",
> ```

nit Consider using `constraint_fn` here to avoid repetition.
___
*[`tutorials/custom_gradients.ipynb` line 295 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTc1lQ8-DSQ59GadRq:-NzTc1lQ8-DSQ59GadRr:b91kkqj) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L295)):*
> ```
>     "        # We are using central differencing here as the `option` argument.\n",
>     "        constraint = constraint_fn(x_value)\n",
>     "        gradient = ComputeNumericalGradient(constraint_fn, x_value, option=NumericalGradientOption(NumericalGradientMethod.kCentral))\n",
> ```

Same as above - missing chain rule?


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 12 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 113 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzElg1u02bM9NAUunNh:-NzTe0tRCfnzYLLgrQ0U:b-8hg15l) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L113)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Sorry to tack on one more thing, but just to ensure we have the shaping correct (e.g. chain rule), ideally we have non-square matrices.

Would it be possible to have one more decision variable, such that the Jacobian will be a 2x3?
</blockquote></details>

Er, scratch that - same number of decision vars, but maybe one more constraint, so that Jacobian is 3x2?


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 12 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 113 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzElg1u02bM9NAUunNh:-NzTlVnJ8zXdzoFGCQ3p:buc26wr) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L113)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Er, scratch that - same number of decision vars, but maybe one more constraint, so that Jacobian is 3x2?
</blockquote></details>

See https://github.com/sea-bass/drake/pull/1 - also includes chain rule stuff


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 6 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 113 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzElg1u02bM9NAUunNh:-NzV9ZA40NLcApOdTehp:b-4yawck) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L113)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

See https://github.com/sea-bass/drake/pull/1 - also includes chain rule stuff
</blockquote></details>

Done (with merge)
___
*[`tutorials/custom_gradients.ipynb` line 141 at r3](https://reviewable.io/reviews/RobotLocomotion/drake/21502#gh-1623146419:-NzV9QqB3j16Ae3Re3Zm:b-il15yg) ([raw file](https://github.com/RobotLocomotion/drake/blob/0189f889760f0b1da6771cdfc5ce697512de857f/tutorials/custom_gradients.ipynb#L141)):*
<details><summary><i>Previously, sea-bass (Sebastian Castro) wrote…</i></summary><blockquote>

```suggestion
    "2x_0 & 2x_1 \\\\ 2(x_0 - 0.5) & 2(x_1 + 0.1)\n",
```

</blockquote></details>

I think this was an accidental publish? Will mark as satisfied for now.
___
*[`tutorials/custom_gradients.ipynb` line 178 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzT_ip2C1xHMLUcVsIy:-NzV9WRr6mg_V5FTMqfB:b-4yawck) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L178)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I believe this might be missing the chain rule? (`ExtractGradient`)

I believe this is currently working because the gradients supplied are currently identity.
However, if the purpose of this tutorial is the correct way to supply gradients to AutoDiff, we should ideally supply the correct value here.
</blockquote></details>

Done (with merge)
___
*[`tutorials/custom_gradients.ipynb` line 199 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTaP8b8yCzno5yWFhH:-NzV9Y4EB2bHsrfc7ko2:b-4yawck) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L199)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Same as above - missing chain rule?
</blockquote></details>

Done (with merge)
___
*[`tutorials/custom_gradients.ipynb` line 270 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTc3cA2pJd7s52MZVE:-NzV9Zns8PNsWEjIqIHQ:b-4yawck) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L270)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Same as above - missing chain rule?
</blockquote></details>

Done (with merge)
___
*[`tutorials/custom_gradients.ipynb` line 284 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTbx8M7c0WWg5tiT8h:-NzV9_Qc4HcysETrwDoH:b-4yawck) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L284)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

nit Consider using `constraint_fn` here to avoid repetition.
</blockquote></details>

Done (with merge)
___
*[`tutorials/custom_gradients.ipynb` line 295 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTc1lQ8-DSQ59GadRq:-NzV9_q42zy5hJ0kwHQN:b-4yawck) ([raw file](https://github.com/RobotLocomotion/drake/blob/2252342409074df682f3b5de20c7c5700b773222/tutorials/custom_gradients.ipynb#L295)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Same as above - missing chain rule?
</blockquote></details>

Done (with merge)


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending remaining items
+@jwnimmer-tri for platform, please!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzV9css7RBsA6U3Slcz:bc2o17x)* status: 6 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)



<!-- Sent from Reviewable.io -->

sea-bass reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 4 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes

___
*[`tutorials/custom_gradients.ipynb` line 61 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzEmpIPE9vkKVVxve50:-NzImyNo9vXgRyK7ecnS:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L61)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

This is exactly right. But I'd like to be super clear here that most people do not need to do this.  We should be clear that even e.g.

```
def cost_expression(x):
    return x[0] + 2.0 * x[1]

prog.AddCost(cost_expression, x)
```

Will work fine -- and support autodiff.  As you say, it would typically only be if you're calling e.g. some black-box method that cannot consume the AutoDiffXd type that you would need to do it.
</blockquote></details>

Done.
___
*[`tutorials/custom_gradients.ipynb` line 199 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzEnjRr9f0AnbA32Uhv:-NzImy9y-lhgF5JeRODm:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/09bf0e5f7e0a3a132aa56e031ff926e1d4bcd76f/tutorials/custom_gradients.ipynb#L199)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Prefer InitializeAutoDiff here, too.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
Great, thanks!  Yes, I'm happy to review but I'll wait until after the output cells are cleared before I review.  It's too cluttered for me to review in the current shape.

> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzVBZjq2hAdYWiMYCrl:b-yqbafp)* status: 4 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)



<!-- Sent from Reviewable.io -->

sea-bass commented:
> Great, thanks! Yes, I'm happy to review but I'll wait until after the output cells are cleared before I review. It's too cluttered for me to review in the current shape.

Fantastic timing -- just did this, plus addressed the remaining minor comments from @EricCousineau-TRI that he didn't do himself.

jwnimmer-tri reviewed (commented):
> Reviewed 1 of 2 files at r2, all commit messages.
> Dismissed @EricCousineau-TRI and @RussTedrake from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzTdnwR3Xe_jWssS_6d:-NzVCu2786cZzfCFlLkQ:bb6un1j) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Before merging, should clear the notebook output.
</blockquote></details>

Done

___


<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@drake-jenkins-bot ok to test

EricCousineau-TRI reviewed (commented):
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)



<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> platform review.

> Reviewed 1 of 4 files at r1, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzVGcsI4SuI_TfsZniU:b-a5n9m0)* status: 7 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`.gitignore` line 34 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVGKDe0c_uo3NAeavq:-NzVGKDe0c_uo3NAeavr:bika8nl) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/.gitignore#L34)):*
> ```ignore
> 
> # Jupyter notebook configuration
> **/.ipynb_checkpoints/
> ```

nit Please revert the changes to this file.  The central purpose of this PR is to add a new tutorial.  Customizing local developers' dotfiles is a fine proposal, but is independent and distinct from the main end-user purpose here, and might require a different set of reviewers / approvals.
___
*[`tutorials/custom_gradients.ipynb` line 10 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVDjHw5amWuCLG-tKS:-NzVDjHw5amWuCLG-tKT:bimdzpw) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L10)):*
> ```
>     "For instructions on how to run these tutorial notebooks, please see the [index](./index.ipynb).\n"
>    ]
>   },
> ```

BTW Somewhere either in the title cell here, or the next cell below, it might help the audience to outline what they are going to learn by going through this document.  Currently it kinda dives right into the lower level details without pointing out where we're headed.
___
*[`tutorials/custom_gradients.ipynb` line 99 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVEQKjADo68bV_wP96:-NzVEQKjADo68bV_wP97:b-ahjykk) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L99)):*
> ```
>     "prog.AddConstraint((x[0] - 0.55)**2 + (x[1] - 0.06)**2 >= 0.1**2)\n",
>     "\n",
>     "result = Solve(prog, np.array([1.0, 1.0]))\n",
> ```

When I run this locally, I get the following answer:

```
Success?  True
SolutionResult.kSolutionFound
x* =  [-9.5e-11  5.0e-01]
cost =  0.7499999997575
```

This doesn't match the `0.48192, 0.13325` answer from the prior figure, but it seems to me like it should have?

Ditto for the custom gradients cell, below.  They all have the `(0, 0.5)` answer.
___
*[`tutorials/custom_gradients.ipynb` line 150 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVFIIX182rrJTmYp3Q:-NzVFIIX182rrJTmYp3R:b6x33ly) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L150)):*
nit

>  NOTE: x_value is a column vector so we specify two indices to avoid warnings

Wouldn't it be more clear to reshape it after the `ExtractValue`, so that we can use 1d indexing pervasively for it?  Especially in the constraint, it would probably help a lot.
___
*[`tutorials/custom_gradients.ipynb` line 171 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVFjhb9-VMi_2GLgf0:-NzVFjhb9-VMi_2GLgf1:b-sfdrmh) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L171)):*
> ```
>     "\n",
>     "        # Compute the constraint and gradient (with chain rule)\n",
>     "        constraint = np.array([\n",
> ```

BTW Here (and possibly also in the cost), I could imagine calling these `y_value` and `y_gradient` for symmetry with `x_value` and `x_gradient`.  That would make it crystal clear that x is the input, y is the output, and each has a value part and a gradient part of similar fates -- they can be unbundled and rebundled using Extract and Initialize.


<!-- Sent from Reviewable.io -->

sea-bass commented:
*[`tutorials/custom_gradients.ipynb` line 99 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVEQKjADo68bV_wP96:-NzVUEqd6jAXhxvwJIkJ:b-klzi9k) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L99)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

When I run this locally, I get the following answer:

```
Success?  True
SolutionResult.kSolutionFound
x* =  [-9.5e-11  5.0e-01]
cost =  0.7499999997575
```

This doesn't match the `0.48192, 0.13325` answer from the prior figure, but it seems to me like it should have?

Ditto for the custom gradients cell, below.  They all have the `(0, 0.5)` answer.
</blockquote></details>

I'm not able to reproduce this -- it seems to run fine consistently on my end with `drake==1.29.0`. What environment are you using?


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 7 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 99 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVEQKjADo68bV_wP96:-NzVUdAkAlpK7ivswC30:bbrii2k) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L99)):*
<details><summary><i>Previously, sea-bass (Sebastian Castro) wrote…</i></summary><blockquote>

I'm not able to reproduce this -- it seems to run fine consistently on my end with `drake==1.29.0`. What environment are you using?
</blockquote></details>

My environment:
- Building from source (using this PR merged up to the latest head of Drake master).
- Ubuntu 22.04 with its default compiler (GCC 11) and default BLAS (netlib).
- Without commercial solvers (no snopt, no mosek, no gurobi).

Atop the cell I just added:
```
import pydrake.common
pydrake.common.configure_logging()
import logging
logging.basicConfig(level=logging.DEBUG)
```

and now the Solve() says

```
DEBUG:drake:solvers::Solve will use IPOPT
```

Very recently on Drake master there were some changes to IPOPT.
\CC @hongkai-dai 


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 7 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 99 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVEQKjADo68bV_wP96:-NzVVu9l4rzQ3P0NJBBi:b6gsk4w) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L99)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

My environment:
- Building from source (using this PR merged up to the latest head of Drake master).
- Ubuntu 22.04 with its default compiler (GCC 11) and default BLAS (netlib).
- Without commercial solvers (no snopt, no mosek, no gurobi).

Atop the cell I just added:
```
import pydrake.common
pydrake.common.configure_logging()
import logging
logging.basicConfig(level=logging.DEBUG)
```

and now the Solve() says

```
DEBUG:drake:solvers::Solve will use IPOPT
```

Very recently on Drake master there were some changes to IPOPT.
\CC @hongkai-dai
</blockquote></details>

If you'd like to test with Drake's nightly build, you can install via pip at https://drake.mit.edu/pip.html#nightly-releases


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 7 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 99 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVEQKjADo68bV_wP96:-NzVWrvS1d8Ia_rbVzIE:boncdz) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L99)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If you'd like to test with Drake's nightly build, you can install via pip at https://drake.mit.edu/pip.html#nightly-releases
</blockquote></details>

I reverted #21499 and #21492 and #21486 and tried again; same result.  So you can probably ignore my ping Hongkai; sorry.

I think it's just luck for which result I see vs you're seeing locally.  If I change the initial guess to `[0.5, 0.5]`, then I get the same answer as you.

WDYT about moving the guess closer to the solution, to make the NLP more likely to find the answer we want?


<!-- Sent from Reviewable.io -->

sea-bass commented:
*[`tutorials/custom_gradients.ipynb` line 10 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVDjHw5amWuCLG-tKS:-NzVdGOO-xPcwtzYuko-:b6ctlsl) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Somewhere either in the title cell here, or the next cell below, it might help the audience to outline what they are going to learn by going through this document.  Currently it kinda dives right into the lower level details without pointing out where we're headed.
</blockquote></details>

Good idea -- added a small preamble.


<!-- Sent from Reviewable.io -->

sea-bass commented:
*[`tutorials/custom_gradients.ipynb` line 99 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVEQKjADo68bV_wP96:-NzVdTCt5oe1I87xtLSA:b-vdowuc) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L99)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I reverted #21499 and #21492 and #21486 and tried again; same result.  So you can probably ignore my ping Hongkai; sorry.

I think it's just luck for which result I see vs you're seeing locally.  If I change the initial guess to `[0.5, 0.5]`, then I get the same answer as you.

WDYT about moving the guess closer to the solution, to make the NLP more likely to find the answer we want?
</blockquote></details>

I'll move the ICs to `[0.5, 0.5]` -- it makes sense to start in the center of the solution space anyhow.

Thanks for digging into this!


<!-- Sent from Reviewable.io -->

sea-bass commented:
*[`tutorials/custom_gradients.ipynb` line 150 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVFIIX182rrJTmYp3Q:-NzVdvT27svFW8bRTaRe:b-c52d61) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L150)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit

>  NOTE: x_value is a column vector so we specify two indices to avoid warnings

Wouldn't it be more clear to reshape it after the `ExtractValue`, so that we can use 1d indexing pervasively for it?  Especially in the constraint, it would probably help a lot.
</blockquote></details>

Nice suggestion! One can indeed throw a `.flatten()` at this problem to squash the NumPy warnings.


<!-- Sent from Reviewable.io -->

sea-bass commented:
*[`.gitignore` line 34 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVGKDe0c_uo3NAeavq:-NzVfQ3uAYwokRUI8Eqv:b-ba7jdc) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/.gitignore#L34)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Please revert the changes to this file.  The central purpose of this PR is to add a new tutorial.  Customizing local developers' dotfiles is a fine proposal, but is independent and distinct from the main end-user purpose here, and might require a different set of reviewers / approvals.
</blockquote></details>

Sounds good to me!


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
LGTM just waiting on a final look from Russ.

> Reviewed 5 of 5 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzVhxgj4Ajpaeo8WJcF:bg8hjd8)* status: 4 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)

___
*[`tutorials/custom_gradients.ipynb` line 171 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVFjhb9-VMi_2GLgf0:-NzVftOE0dKUOaFh8YIC:b-xukos2) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L171)):*
<details><summary><i>Previously, sea-bass (Sebastian Castro) wrote…</i></summary><blockquote>

I'm not sure I agree here -- I think since the decision variables are named `x = [x0, x1]`, expressing things in terms of `x` is clearer than introducing another letter.

Plus for the chain rule expressions, it reads as `grad(f(x)) = g(x) * grad(x)`, where `g(x)` is your analytically provided gradient.
</blockquote></details>

Maybe I wasn't clear?

My proposal was to "name" the return value of `constraint_expression` to be `y`, which matches with C++ [output argument](https://drake.mit.edu/doxygen_cxx/classdrake_1_1solvers_1_1_evaluator_base.html#a515e0d153b83bd7133f6a4bbf5e4534e) -- C++ uses an output argument instead of a return value to reduce heap allocations.

Something along the lines of:

```py
y_value = ... math on x ...
y_grad = ... math on x ... @ x_grad
y = InitializeAutoDiff(y_value, y_grad)
return y
```

This hopefully highlights the fact that every array-with-partials can be modeled as a pairing of the value array and the partials array, and we can split and recombine (aka zip and unzip) those as needed.

It's OK if you think the existing spelling is better, but at least I wanted to be sure I was clear.
___
*[`tutorials/custom_gradients.ipynb` line 176 at r10](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVhPcQ42dC8BfOsya9:-NzVhPcR6AUfE21I9gnW:b-uh0y3a) ([raw file](https://github.com/RobotLocomotion/drake/blob/e0b6866bf9845816f1bc50bbf34b0d8cd0af414b/tutorials/custom_gradients.ipynb#L176)):*
> ```
>     "        # Compute the constraint and gradient (with chain rule)\n",
>     "        constraint = np.array([\n",
>     "            x_value[0, 0]**2 + x_value[1, 0]**2,\n",
> ```

BTW Did you want to flatten the `x_value` throughout the entire tutorial?  This one didn't change.


<!-- Sent from Reviewable.io -->

sea-bass reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 4 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes

___
*[`tutorials/custom_gradients.ipynb` line 171 at r7](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVFjhb9-VMi_2GLgf0:-NzVoeEX3gDkwCcOZVNm:b7b3hdv) ([raw file](https://github.com/RobotLocomotion/drake/blob/ae9634f379029e0523b123a8cdbe13ad230b65ee/tutorials/custom_gradients.ipynb#L171)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Maybe I wasn't clear?

My proposal was to "name" the return value of `constraint_expression` to be `y`, which matches with C++ [output argument](https://drake.mit.edu/doxygen_cxx/classdrake_1_1solvers_1_1_evaluator_base.html#a515e0d153b83bd7133f6a4bbf5e4534e) -- C++ uses an output argument instead of a return value to reduce heap allocations.

Something along the lines of:

```py
y_value = ... math on x ...
y_grad = ... math on x ... @ x_grad
y = InitializeAutoDiff(y_value, y_grad)
return y
```

This hopefully highlights the fact that every array-with-partials can be modeled as a pairing of the value array and the partials array, and we can split and recombine (aka zip and unzip) those as needed.

It's OK if you think the existing spelling is better, but at least I wanted to be sure I was clear.
</blockquote></details>

Got it -- that makes sense!
___
*[`tutorials/custom_gradients.ipynb` line 176 at r10](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-NzVhPcQ42dC8BfOsya9:-NzVogPv1oPGpezG6XZb:b-vt63tn) ([raw file](https://github.com/RobotLocomotion/drake/blob/e0b6866bf9845816f1bc50bbf34b0d8cd0af414b/tutorials/custom_gradients.ipynb#L176)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Did you want to flatten the `x_value` throughout the entire tutorial?  This one didn't change.
</blockquote></details>

Whoops. Fixed.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> Reviewed 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502)* status: 2 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on @sea-bass)



<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
Looks great.  We'll give Russ the weekend in case he has time to take a last look, otherwise I'll plan to merge on Monday.

> Reviewed 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NznbKp7CpBoLEGMJab7:b-wypbhv)* status: 2 unresolved discussions, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @sea-bass)



<!-- Sent from Reviewable.io -->

RussTedrake reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  thanks!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/21502#-:-NzoPqzNBVVLopLno9Np:btapcqv)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @sea-bass)



<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Thanks for this!

FYI It's reached Drake's public website now, as part of the v1.30.0 release:
https://deepnote.com/workspace/Drake-0b3b2c53-a7ad-441b-80f8-bf8350752305/project/Tutorials-2b4fc509-aef2-417d-a40d-6071dfed9199/notebook/custom_gradients-7fc33ebf63674e9faa9f8aa0b98d1091
